code/ndarray.c

```diff
@@ -912,12 +912,38 @@ STATIC mp_obj_t ndarray_make_new_core(const mp_obj_type_t *type, size_t n_args,
     return MP_OBJ_FROM_PTR(self);
 }
 
-mp_obj_t ndarray_array_constructor(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
-    // array constructor for ndarray, equivalent to numpy.array(...)
-    return ndarray_make_new_core(&ulab_ndarray_type, n_args, kw_args->used, pos_args, kw_args);
-}
 MP_DEFINE_CONST_FUN_OBJ_KW(ndarray_array_constructor_obj, 1, ndarray_array_constructor);
 
+STATIC mp_obj_t ndarray_make_new_core(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args, mp_map_t *kw_args) {
+    (void) type;
+    mp_arg_check_num(n_args, n_kw, 1, 2, true);
+    uint8_t dtype = ndarray_init_helper(n_args, args, kw_args);
+
+    mp_obj_t mp_shape = args[0];
+    mp_obj_t mp_ndim_maybe = mp_obj_len_maybe(mp_shape);
+    // single-number shapes "x" are interpreted the same as "(x,)"
+    mp_int_t ndim;
+    size_t shape[ULAB_MAX_DIMS];
+    
+    if (mp_ndim_maybe == MP_OBJ_NULL) {
```

Comment:
This will fail with `ndarray(-12.3)`. You can make it safe by explicitly checking for the input type: https://github.com/v923z/micropython-ulab/blob/ab964b94be79371bdb263f451793a1b3d943dfa1/code/ulab_create.c#L29